test/Dialect/Moore/canonicalizers.mlir

```diff
@@ -10,32 +10,168 @@ func.func @Casts(%arg0: !moore.i1) -> (!moore.i1, !moore.i1) {
   return %0, %1 : !moore.i1, !moore.i1
 }
 
-// CHECK-LABEL: moore.module @SingleAssign
-moore.module @SingleAssign() {
+// CHECK-LABEL: moore.module @OptimizeUniquelyAssignedVars
+moore.module @OptimizeUniquelyAssignedVars(in %u: !moore.i42, in %v: !moore.i42, in %w: !moore.i42) {
+  // Unique continuous assignments to variables should remove the `ref<T>`
+  // indirection and instead directly propagate the assigned value to readers.
+  // CHECK-NOT: moore.assign
   // CHECK-NOT: moore.variable
-  // CHECK: %a = moore.assigned_variable %0 : <i32>
-  %a = moore.variable : <i32>
-  // CHECK: %0 = moore.constant 32 : i32
-  %0 = moore.constant 32 : i32
-  // CHECK: moore.assign %a, %0 : i32
-  moore.assign %a, %0 : i32
-  moore.output
+  // CHECK: %a = moore.assigned_variable %u : i42
+  // CHECK: dbg.variable "a", %a : !moore.i42
+  moore.assign %a, %u : i42
+  %a = moore.variable : <i42>
+  %3 = moore.read %a : <i42>
+  dbg.variable "a", %3 : !moore.i42
+
+  // Continuous assignments to variables should override the initial value.
+  // CHECK-NOT: moore.assign
+  // CHECK-NOT: moore.constant 9001
+  // CHECK-NOT: moore.variable
+  // CHECK: %b = moore.assigned_variable %v : i42
+  // CHECK: dbg.variable "b", %b : !moore.i42
+  moore.assign %b, %v : i42
+  %0 = moore.constant 9001 : i42
+  %b = moore.variable %0 : <i42>
+  %4 = moore.read %b : <i42>
+  dbg.variable "b", %4 : !moore.i42
+
+  // Unique continuous assignments to nets should remove the `ref<T>`
+  // indirection and instead directly propagate the assigned value to readers.
+  // CHECK-NOT: moore.assign
+  // CHECK-NOT: moore.net wire
+  // CHECK: %c = moore.assigned_variable %w : i42
+  // CHECK: dbg.variable "c", %c : !moore.i42
+  moore.assign %c, %w : i42
+  %c = moore.net wire : <i42>
```

Comment:
Sorry, but a couple of dumb SV questions:

If we can replace a `net wire` with an `assigned_variable` can we also replace it with just a `variable`? Is this only allowed for uniquely driven wires? Otherwise, why are we distinguishing between wire and variable?

Also, for the cases where we have an initial value given to the variable or net operation, wouldn't that be noticeable in a simulation trace because at `t=0s` it would be the initial value and at time `t=0s,0d,1e`  it would be the driven value? Do we only care about the real time steps?

Reply:
I think in case there is a single unique driver to the net, the net is actually equivalent to the variable. If I recall correctly, the differences start when you have multiple drivers to the same net or variable, where a net will potentially have different driving conflict resolution compared to the variable. My guess is that we'll be able to make the conflict resolution explicit in an operation at some point, and then have optimization passes at the Moore level where we replace wires with simpler variables, or plain values directly, whenever the fancy features of nets aren't used. One example would be inout ports, or the nets we conjure up to model instance outputs driving values. A lot of these should turn out to be simplifiable to variables or even to plain values.

Good point about the assignments and delays. IIRC the SV spec allows for continuous assignments to be interpreted as aliases, such that they don't introduce a delta time delay. Which in this case would mean that there wouldn't be any difference observable at `t=0s`. I remember that some commercial tools make heavy use of this to reduce the number of iterations on events 🤔. I'll have to do some more digging in the SV spec on what the _exact_ semantics are, but I think most EDA tools apply optimizations that remove delta time steps.

Reply:
Found the following in IEEE 1800-2017 "4.9.1 Continuous assignment":
> A continuous assignment process is also evaluated at time zero in order to propagate constant values. This includes implicit continuous assignments inferred from port connections (see 3.9.6).